extended-utxo-spec/extended-utxo-specification.tex

```diff
@@ -0,0 +1,1212 @@
+  \end{itemize}
+  \caption{Modified \valuemap{} type for XUTXO-3}
+  \label{fig:xutxo-3-types}
+\end{ruledfigure}
+
+\noindent Note that the comparison and addition operations for maps defined in
+Figure~\ref{fig:basic-notation} carry over to the type \tokenmap.
+This allows us to define XUTXO-3 by exactly the same types and
+validity rules as XUTXO-2, simply replacing \valuemap{} by
+\tokenmap{} everywhere.
+
+\medskip \noindent\red{I'm slightly confused about what we want to do here.  In
+  Jann's multicurrency proposal, we have a type \s{Quantity} which is
+  used to measure quantities of tokens, but it's left open-ended
+  exactly what \s{Quantity} should be.  The problem is that if we're
+  following the convention of destroying things by forging
+  negative amounts, then we have to know what ``negative'' means, and
+  I think you need at least an ordered abelian group for that.  Will
+  anyone ever use anything that isn't \Z?}
+
+\subsection{Implications}
+This model allows us to implement fungible (normal) and non-fungible
+token currencies, as well as ``mixed states'':
+\begin{itemize}
+\item Fungible token currencies are implemented by only ever issuing
+  \s{Value}s of a single \s{Token}.
+\item Non-fungible token currencies are implemented by only ever
+  issuing single \s{Value}s of many unique \s{Token}s.
+\item Note that there is nothing in this model which enforces
+  uniqueness: having multiples of a single \s{Token} merely means that
+  those can be used fungibly. If a currency wants to make sure it only
+  issues unique tokens it must track this itself.  These ``mixed'' token
+  currencies can have many \s{Token}s, but these can have more than unit
+  \s{Value}s in circulation.  These can be useful to model distinct
+  categories of thing where there are fungible quantities within
+  those, for example share classes. \red{Eh?}
+\end{itemize}
+
+\paragraph{Performance.}
+\red{I'm just making random stuff up here.}
+The UTXO-2 and UTXO-3 models will lose some efficiency in comparison
+to the UTXO-1 model, simply because the data structures are more
+complicated.  This would even apply to transactions which only involve
+to native currency, since it would be necessary to check whether the
+\valuemap{} or \tokenmap{} contains anything that needs to be processed.  If this is
+a concern then one could implement a model with two types of
+transaction, essentially just the disjoint union of the XUTXO-1 and
+XUTXO-2 or XUTXO-3 transaction types. A simple case distinction at the
+start of a transaction could then select either a fast
+native-currency-only computation or a slower multicurrency computation.
+This would be harder to maintain though.
+
+\medskip
+A more significant cost may be that we can no longer use \verb|{-# UNPACK #-}|
+when our Value type stops being a simple combination of wrappers
+and products around primitives, but this is again an issue with any
+multi-currency proposal.
```

Comment:
I wonder if the argument that most of the time it takes to validate a transaction is spent doing cryptography things applies here as well.